### PR TITLE
BREAKING(async): simplify `deadline()` logic, remove `DeadlineError` and improve errors

### PR DESCRIPTION
**Pull request by iuioiua**
### What's changed

This PR:
1. Ensures that all thrown/rejected errors within [`@std/async`](https://jsr.io/@std/async) APIs are [`DOMException`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException)s, with the appropriate [`AbortError`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException#aborterror) or [`TimeoutError`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException#timeouterror) error names.
2. Simplifies [`deadline()`](https://jsr.io/@std/async/doc/~/deadline) by relying on [`abortable()`](https://jsr.io/@std/async/doc/~/abortable).
3. Removes `DeadlineError` for `deadline()`, as it's no longer needed.
4. Clarifies the error message when one of these functions is called and the `AbortSignal` is already aborted.

### Why was this change been made?

These changes were made to:
1. Ensure this package's APIs are consistent in how they throw errors.
2. Reduce the number of moving parts.
3. Reduce the need to import an additional custom error, `DeadlineError`.
4. Improve the clarity of errors and therefore help with troubleshooting.

### Migration guide

If using `deadline()` compare the error with `DOMException` instead of `DeadlineError` when handling errors.

```diff
- import { deadline, type DeadlineError } from "@std/async";
+ import { deadline } from "@std/async";

  try {
    await deadline(promise, 1_000);
  } catch (error) {
-   if (error instanceof DeadlineError && error.message === "Deadline") {
+   if (error instanceof DOMException && error.name === "TimeoutError") {
      // Handle error
    }
  }
```

### Related

Depends on https://github.com/denoland/deno/pull/23842

### CODE REVIEW

**Comment by kt3k** [comment]
> The main consequences of these changes are that the error type has changed from DeadlineError to DOMException and the error message from Deadline to The signal has been aborted. Deadline error has been removed as it's no longer used.

I don't understand why we want to make this change. Currently the user can distinguish whether the promise is aborted by signal or aborted by deadline by catching the error. Now they can't do it.

**Comment by iuioiua** [comment]
You can compare using `error.name === "AbortError"` or `error.name === "TimeoutError"`, similar to `AbortSignal.timeout()`. I've clarified my initial comment.

**Comment by kt3k** [comment]
> You can compare using error.name === "AbortError" or error.name === "TimeoutError", similar to AbortSignal.timeout().

Oh, this sounds nice.

> I've clarified my initial comment. Perhaps, we can update the documentation as well.

Sounds a good idea. Can you update the doc examples and test case to demonstrate this?

File: async/deadline_test.ts [hunk 1]
```diff
@@ -14,15 +14,19 @@ Deno.test("deadline() returns fulfilled promise", async () => {
...
-  await assertRejects(async () => {
-    await deadline(p, 100);
-  }, DeadlineError);
+  await assertRejects(
+    async () => {
+      await deadline(p, 100);
+    },
+    DOMException,
+    "Signal timed out.",
+  );
```
Comment (kt3k) [inline async/deadline_test.ts:29]: Maybe we should also assert `err.name === "TimeoutError"`?

File: async/deadline_test.ts [hunk 2]
```diff
@@ -77,8 +87,13 @@ Deno.test("deadline() handles already aborted signal", async () => {
...
   abort.abort();
-  await assertRejects(async () => {
-    await deadline(p, 100, { signal: abort.signal });
-  }, DeadlineError);
+  const error = await assertRejects(
+    async () => {
+      await deadline(p, 100, { signal: abort.signal });
+    },
+    DOMException,
+    "The signal has been aborted",
```
Comment (iuioiua) [inline async/deadline_test.ts:95]: Hm... The discrepancy between the error message in this test and the one above seems odd.

**Comment by iuioiua** [comment]
I've decided to revert the use of `signal.throwIfAborted()`, as it throws the reason, which is a string, instead of an error with the message as the reason.

**Comment by iuioiua** [comment]
I've tweaked the direction of this PR and re-written the descript. @kt3k, please take a look once more.

**Comment by 0f-0b** [comment]
> I've decided to revert the use of `signal.throwIfAborted()`, as it throws the reason, which is a string, instead of an error with the message as the reason.

I think it's better to always just throw the reason itself, to be consistent with Web APIs like `fetch` and [`navigator.locks.request`](https://w3c.github.io/web-locks/#dom-lockmanager-request).

**Comment by iuioiua** [comment]
> I think it's better to always just throw the reason itself, to be consistent with Web APIs like `fetch` and [`navigator.locks.request`](https://w3c.github.io/web-locks/#dom-lockmanager-request).

This code does, wrapped in a `DOMException`. Do you see any points where it doesn't?

File: async/mod.ts [hunk 3]
```diff
@@ -1,5 +1,9 @@
 // Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
-// This module is browser compatible.
+// TODO(iuioiua): Add web-compatible declaration once TypeScript 5.5 is released
+// and in the Deno runtime. See https://github.com/microsoft/TypeScript/pull/58211
+//
+// Note: this code is still compatible with recent
+// web browsers. See https://caniuse.com/?search=AbortSignal.any
```
Comment (kt3k) [inline async/mod.ts:6]: This seems no longer necessary?

File: async/delay.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ export interface DelayOptions {
 /**
  * Resolve a {@linkcode Promise} after a given amount of milliseconds.
  *
+ * @throws {DOMException} If the optional signal is already aborted.
```
Comment (kt3k) [inline async/delay.ts:18]:
```suggestion
 * @throws {DOMException} If the optional signal is aborted before the delay duration.
```

File: async/mod.ts (hunk 3 above)
Comment (iuioiua) [reply async/mod.ts:6]: It is necessary because `deadline()` uses `AbortSignal.any()`.

**Comment by 0f-0b** [comment]
> This code does, wrapped in a `DOMException`. Do you see any spots where it doesn't?

Here the reason (which is usually [already an `"AbortError"` `DOMException`](https://developer.mozilla.org/docs/Web/API/AbortController/abort#reason)) is implicitly converted to a string and used as the message of another exception. That is no longer the reason itself.

**Comment by kt3k** [comment]
Maybe `ac.abort()` method suppose the reason should be an error object instead of string?

**Comment by iuioiua** [comment]
Oh, I see. Thank you. I've made some adjustments, which made things even cleaner. PTAL.

File: async/abortable.ts [hunk 5]
```diff
@@ -77,11 +125,9 @@ function abortablePromise<T>(
   p: Promise<T>,
   signal: AbortSignal,
 ): Promise<T> {
-  if (signal.aborted) {
-    return Promise.reject(createAbortError(signal.reason));
-  }
+  if (signal.aborted) return Promise.reject(signal.reason);
```
Comment (iuioiua) [inline async/abortable.ts:128]: On 2nd thought. Maybe this should throw. Thoughts?

File: async/abortable.ts [hunk 6]
```diff
@@ -1,61 +1,109 @@
...
+ * ```ts
+ * import { abortable, delay } from "@std/async";
+ * import { assertRejects, assertEquals } from "@std/assert";
  *
- * const p = delay(1000);
- * const c = new AbortController();
- * setTimeout(() => c.abort(), 100);
+ * const promise = delay(1_000);
+ * const controller = new AbortController();
+ * controller.abort(new Error("This is my reason"));
```
Comment (kt3k) [inline async/abortable.ts:36]: 👍

File: async/mod.ts (hunk 3 above)
Comment (kt3k) [reply async/mod.ts:6]: Ah, ok

File: async/abortable.ts (hunk 5 above)
Comment (kt3k) [reply async/abortable.ts:128]: This looks fine to me

**Review by kt3k — APPROVED** [review]
LGTM

I think the removal of wrapping of abort reason is a good move towards Web standard alignment. Thanks @0f-0b for the ideas!

**Pull request merged**